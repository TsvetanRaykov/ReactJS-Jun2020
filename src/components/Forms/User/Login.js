import React, { useState, useContext } from 'react'
import {
	Typography,
	Paper,
	Avatar,
	Button,
	FormControl,
	Input,
	InputLabel,
} from '@material-ui/core'
import { Link, withRouter } from 'react-router-dom'
import userService from '../../../services/userService'
import Loader from '../../Loader'
import Context from '../../../Context'
import { withSnackbar } from 'notistack'

const SignIn = (props) => {
	const { classes } = props
	const { updateUser } = useContext(Context)

	const [email, setEmail] = useState('')
	const [password, setPassword] = useState('')
	const [loading, setLoading] = useState(false)

	return loading ? (
		<Loader />
	) : (
		<main className={classes.main}>
			<Paper className={classes.paper}>
				<Avatar className={classes.avatar} src='/logo.jpg' />
				<Typography component='h1' variant='h5'>
					Sign in
				</Typography>
				<form
					className={classes.form}
					onSubmit={(e) => e.preventDefault() && false}
				>
					<FormControl margin='normal' required fullWidth>
						<InputLabel htmlFor='email'>Email Address</InputLabel>
						<Input
							id='email'
							name='email'
							autoComplete='off'
							autoFocus
							value={email}
							onChange={(e) => setEmail(e.target.value)}
						/>
					</FormControl>
					<FormControl margin='normal' required fullWidth>
						<InputLabel htmlFor='password'>Password</InputLabel>
						<Input
							name='password'
							type='password'
							id='password'
							autoComplete='off'
							value={password}
							onChange={(e) => setPassword(e.target.value)}
						/>
					</FormControl>
					<Button
						type='submit'
						fullWidth
						variant='contained'
						color='primary'
						onClick={login}
						className={classes.submit}
					>
						Sign in
					</Button>
					<Button
						type='submit'
						fullWidth
						variant='contained'
						color='secondary'
						component={Link}
						to='/register'
						className={classes.submit}
					>
						Register
					</Button>
				</form>
			</Paper>
		</main>
	)

	async function login() {
		setLoading(true)
		userService
			.login(email, password)
			.then(() => {
				updateUser(userService.getCurrentUser())
				setLoading(false)
				props.history.replace('/dashboard')
			})
			.catch(() => {
				props.enqueueSnackbar('Login failed', {
					variant: 'error',
				})
			})
			.finally(() => {
				setLoading(false)
			})
	}
}
export { SignIn }
export default withSnackbar(withRouter(SignIn))
