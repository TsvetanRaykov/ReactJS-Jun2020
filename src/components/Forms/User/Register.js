import React, { useState } from 'react'
import {
	Typography,
	Paper,
	Avatar,
	Button,
	FormControl,
	Input,
	InputLabel,
	FormHelperText,
} from '@material-ui/core'
import { Link, withRouter } from 'react-router-dom'
import userService from '../../../services/userService'
import { withSnackbar } from 'notistack'

function Register(props) {
	const { classes } = props

	const [name, setName] = useState('')
	const [email, setEmail] = useState('')
	const [password, setPassword] = useState('')
	const [repassword, setRepassword] = useState('')

	const [validators, setValidators] = useState({
		name,
		email,
		password,
		repassword,
	})

	const validEmailRegex = RegExp(
		/^(([^<>()[\].,;:\s@"]+(\.[^<>()[\].,;:\s@"]+)*)|(".+"))@(([^<>()[\].,;:\s@"]+\.)+[^<>()[\].,;:\s@"]{2,})$/i
	)

	return (
		<main className={classes.main}>
			<Paper className={classes.paper}>
				<Avatar className={classes.avatar} src='/logo.jpg'></Avatar>
				<Typography component='h1' variant='h5'>
					Register Account
				</Typography>
				<form
					className={classes.form}
					onSubmit={(e) => e.preventDefault() && false}
				>
					<FormControl margin='normal' required fullWidth>
						<InputLabel htmlFor='name'>Name</InputLabel>
						<Input
							id='name'
							name='name'
							error={!!validators.name}
							autoComplete='off'
							autoFocus
							value={name}
							onChange={(e) => setName(e.target.value)}
						/>
						<FormHelperText id='name-helper-text'>
							{validators.name}
						</FormHelperText>
					</FormControl>
					<FormControl margin='normal' required fullWidth>
						<InputLabel htmlFor='email'>Email Address</InputLabel>
						<Input
							id='email'
							name='email'
							error={!!validators.email}
							autoComplete='off'
							value={email}
							onChange={(e) => setEmail(e.target.value)}
						/>
						<FormHelperText id='name-helper-text'>
							{validators.email}
						</FormHelperText>
					</FormControl>
					<FormControl margin='normal' required fullWidth>
						<InputLabel htmlFor='password'>Password</InputLabel>
						<Input
							name='password'
							type='password'
							id='password'
							autoComplete='off'
							value={password}
							error={!!validators.password}
							onChange={(e) => setPassword(e.target.value)}
						/>
						<FormHelperText id='name-helper-text'>
							{validators.password}
						</FormHelperText>
					</FormControl>
					<FormControl margin='normal' required fullWidth>
						<InputLabel htmlFor='password'>Confirm Password</InputLabel>
						<Input
							name='repassword'
							type='password'
							id='repassword'
							autoComplete='off'
							value={repassword}
							error={!!validators.repassword}
							onChange={(e) => setRepassword(e.target.value)}
						/>
						<FormHelperText id='name-helper-text'>
							{validators.repassword}
						</FormHelperText>
					</FormControl>
					<Button
						type='submit'
						fullWidth
						variant='contained'
						color='primary'
						onClick={onRegister}
						className={classes.submit}
					>
						Register
					</Button>

					<Button
						type='submit'
						fullWidth
						variant='contained'
						color='secondary'
						component={Link}
						to='/login'
						className={classes.submit}
					>
						Login
					</Button>
				</form>
			</Paper>
		</main>
	)

	function validateForm() {
		const errors = { name, email, password, repassword }
		errors.name =
			!name || name.length < 3 ? 'Name must be at least 3 characters.' : ''

		errors.email = validEmailRegex.test(email)
			? ''
			: 'Email address is invalid.'

		errors.password =
			password.length < 6 ? 'Password must be at least 6 characters.' : ''

		errors.repassword =
			!errors.password && password !== repassword
				? 'Passwords does not match.'
				: ''

		setValidators(errors)

		return !Object.entries(errors).find(([, v]) => v.length > 0)
	}

	async function onRegister() {
		if (validateForm()) {
			try {
				await userService.register(name, email, password)
				props.history.replace('/dashboard')
			} catch (e) {
				const message =
					e?.code === 'auth/email-already-in-use'
						? 'Email is already registered'
						: 'Registraion failed'
				props.enqueueSnackbar(message, {
					variant: 'error',
				})
			}
		}
	}
}

export { Register }
export default withSnackbar(withRouter(Register))
