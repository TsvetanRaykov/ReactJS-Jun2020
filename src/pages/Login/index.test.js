import React from 'react'
import renderer from 'react-test-renderer'
import TestWrapper from '../../utils/test-wrapper'
import { SignIn as Login } from '../../components/Forms/User/Login'

jest.mock('react', () => ({
	...jest.requireActual('react'),
	useLayoutEffect: jest.requireActual('react').useEffect,
}))

describe('<Login />', () => {
	it('Should match the snapshot', () => {
		const wrapper = renderer.create(
			<TestWrapper>
				<Login classes={{ main: {} }} />
			</TestWrapper>
		)
		expect(wrapper).toMatchSnapshot()
	})
})
