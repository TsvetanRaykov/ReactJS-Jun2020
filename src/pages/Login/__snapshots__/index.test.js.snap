// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Login /> Should match the snapshot 1`] = `
<main
  className={Object {}}
>
  <div
    className="MuiPaper-root MuiPaper-elevation1 MuiPaper-rounded"
  >
    <div
      className="MuiAvatar-root MuiAvatar-circular"
    >
      <img
        className="MuiAvatar-img"
        src="/logo.jpg"
      />
    </div>
    <h1
      className="MuiTypography-root MuiTypography-h5"
    >
      Sign in
    </h1>
    <form
      onSubmit={[Function]}
    >
      <div
        className="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
          data-shrink={false}
          htmlFor="email"
        >
          Email Address
          <span
            aria-hidden={true}
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoComplete="off"
            autoFocus={true}
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            id="email"
            name="email"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="text"
            value=""
          />
        </div>
      </div>
      <div
        className="MuiFormControl-root MuiFormControl-marginNormal MuiFormControl-fullWidth"
      >
        <label
          className="MuiFormLabel-root MuiInputLabel-root MuiInputLabel-formControl MuiInputLabel-animated Mui-required Mui-required"
          data-shrink={false}
          htmlFor="password"
        >
          Password
          <span
            aria-hidden={true}
            className="MuiFormLabel-asterisk MuiInputLabel-asterisk"
          >
             
            *
          </span>
        </label>
        <div
          className="MuiInputBase-root MuiInput-root MuiInput-underline MuiInputBase-formControl MuiInput-formControl"
          onClick={[Function]}
        >
          <input
            aria-invalid={false}
            autoComplete="off"
            className="MuiInputBase-input MuiInput-input"
            disabled={false}
            id="password"
            name="password"
            onAnimationStart={[Function]}
            onBlur={[Function]}
            onChange={[Function]}
            onFocus={[Function]}
            required={true}
            type="password"
            value=""
          />
        </div>
      </div>
      <button
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedPrimary MuiButton-fullWidth"
        disabled={false}
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="submit"
      >
        <span
          className="MuiButton-label"
        >
          Sign in
        </span>
      </button>
      <a
        aria-disabled={false}
        className="MuiButtonBase-root MuiButton-root MuiButton-contained MuiButton-containedSecondary MuiButton-fullWidth"
        href="/register"
        onBlur={[Function]}
        onClick={[Function]}
        onDragLeave={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        role="button"
        tabIndex={0}
      >
        <span
          className="MuiButton-label"
        >
          Register
        </span>
      </a>
    </form>
  </div>
</main>
`;
